src/basic/smack-util.c

```diff
@@ -45,84 +45,162 @@ bool mac_smack_use(void) {
+        char l[NAME_MAX] = "", *ll = NULL;
+        ssize_t s = 0;
 
         assert(path);
+        assert(attr);
+        assert(label);
 
 #ifdef HAVE_SMACK
         if (!mac_smack_use())
                 return 0;
 
-        if (label)
-                r = lsetxattr(path, "security.SMACK64", label, strlen(label), 0);
-        else
-                r = lremovexattr(path, "security.SMACK64");
-        if (r < 0)
+        s = getxattr(path, attr, l, NAME_MAX);
+        if (s < 0)
                 return -errno;
+
+        ll = strdup(l);
+        if (!ll)
+                return -ENOMEM;
+
+        *label = ll;
 #endif
 
-        return r;
+        return s;
 }
 
-int mac_smack_apply_fd(int fd, const char *label) {
-        int r = 0;
+static int mac_get_xattr_fd(int fd, const char *attr, const char **label) {
+        char l[NAME_MAX] = "", *ll = NULL;
+        ssize_t s = 0;
 
         assert(fd >= 0);
+        assert(attr);
+        assert(label);
 
 #ifdef HAVE_SMACK
         if (!mac_smack_use())
                 return 0;
 
-        if (label)
-                r = fsetxattr(fd, "security.SMACK64", label, strlen(label), 0);
-        else
-                r = fremovexattr(fd, "security.SMACK64");
-        if (r < 0)
+        s = fgetxattr(fd, attr, l, NAME_MAX);
+        if (s < 0)
                 return -errno;
+
+        ll = strdup(l);
+        if (!ll)
+                return -ENOMEM;
+
+        *label = ll;
 #endif
 
-        return r;
+        return s;
 }
 
-int mac_smack_apply_ip_out_fd(int fd, const char *label) {
-        int r = 0;
+static int mac_set_xattr(const char *path, const char *attr, const char *label) {
+        int r;
 
-        assert(fd >= 0);
+        assert(path);
+        assert(attr);
 
 #ifdef HAVE_SMACK
         if (!mac_smack_use())
                 return 0;
 
         if (label)
-                r = fsetxattr(fd, "security.SMACK64IPOUT", label, strlen(label), 0);
+                r = lsetxattr(path, attr, label, strlen(label), 0);
         else
-                r = fremovexattr(fd, "security.SMACK64IPOUT");
+                r = lremovexattr(path, attr);
         if (r < 0)
                 return -errno;
 #endif
 
-        return r;
+        return 0;
 }
 
-int mac_smack_apply_ip_in_fd(int fd, const char *label) {
-        int r = 0;
+static int mac_set_xattr_fd(int fd, const char *attr, const char *label) {
+        int r;
 
         assert(fd >= 0);
+        assert(attr);
 
 #ifdef HAVE_SMACK
         if (!mac_smack_use())
                 return 0;
 
         if (label)
-                r = fsetxattr(fd, "security.SMACK64IPIN", label, strlen(label), 0);
+                r = fsetxattr(fd, attr, label, strlen(label), 0);
         else
-                r = fremovexattr(fd, "security.SMACK64IPIN");
+                r = fremovexattr(fd, attr);
         if (r < 0)
                 return -errno;
 #endif
 
-        return r;
+        return 0;
+}
+
+int mac_smack_read(const char *path, const char **label) {
```

Comment:
The asserts still need to live in these functions. Otherwise, the message that emerges doesn't reflect the function which was called.